src/internal/connector/graph/service.go

```diff
@@ -366,3 +380,56 @@ func (middleware RetryHandler) Intercept(
 
 	return response, nil
 }
+
+// ThrottleControlMiddleware is used to ensure we don't overstep 10k-per-10-min
+// request limits.  Currently a rudimentary implementation, and in need of
+// configuration refinement.
+// TODO: doesn't actually save us until this tracker is shared across all http
+// clients as a singleton.  But this gets the point across.
+type ThrottleControlMiddleware struct {
```

Comment:
@ryanfkeepers - can you take a look at https://pkg.go.dev/golang.org/x/time/rate#Limiter (there are a few other ones as well) - anything that implements the token-bucket algorithm

Reply:
if possible, it may be wise to prioritize using one that has something like a fifo policy for which request gets to go when credits are replenished. That will help let us tighten the context/http deadlines down the road as we can make some assumptions about things moving out of the "waiting" group

Reply:
> fifo policy

Limiter doesn't have a strict queue, but it doesn't have capacity for resource contention/starvation either.  Each request is given a single delay, and does not re-evaluate resources at the end of the delay.  This gives us, in effect, fifo behavior.

Reply:
alright, that should work then. Mostly I was worried about the possibility to have one request keep getting put off and then fail with a context deadline or HTTP timeout. If there's at least some mechanism to keep that from happening it should be sufficient. Thanks for digging into this more to figure it out

Reply:
It's a good concern to guarantee, for sure.  Happy to look into it.